File: sources/svm/main/svm-test.cpp
// Copyright 2015, 2016, 2017 Ingo Steinwart
//
// This file is part of liquidSVM.
//
// liquidSVM is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as 
// published by the Free Software Foundation, either version 3 of the 
// License, or (at your option) any later version.
//
// liquidSVM is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public License
// along with liquidSVM. If not, see <http://www.gnu.org/licenses/>.


#include <string.h>
#include <cstdlib>


#include "sources/shared/system_support/timing.h"
#include "sources/shared/system_support/parallel_control.h"
#include "sources/shared/basic_functions/flush_print.h"
#include "sources/shared/basic_functions/basic_file_functions.h"
#include "sources/shared/basic_types/dataset_info.h"
#include "sources/shared/training_validation/working_set_manager.h"


#include "sources/svm/training_validation/svm_train_val_info.h"
#include "sources/svm/decision_function/svm_decision_function_manager.h"
#include "sources/svm/command_line/command_line_parser_svm_test.h"




#include "sources/svm/training_validation/svm_manager.h"


//**********************************************************************************************************************************

int main(int argc, char** argv)
{
	Tcommand_line_parser_svm_test command_line_parser;

	Tdataset training_set;
	Tdataset test_set;

	Tdataset_info testset_info;

	Tsvm_full_test_info test_info;
	Tsvm_manager svm_manager;
	
	vector <double> predictions;

	double detection_rate;
	double false_alarm_rate;
	double full_time;

	FILE* fplogwrite;
	FILE* fpresultwrite;

	unsigned i;
	unsigned j;
	unsigned task_offset;


//--------- Read command line ---------------------------------------------------------------------------- 

	full_time = get_wall_time_difference();
	command_line_parser.setup(argc, argv);
	command_line_parser.parse(true);
	
	
//------- Load training set, the solutions & their performance, and the test data ------------------------

	test_info.file_time = get_process_time_difference();
	training_set.read_from_file(command_line_parser.train_filename);
	test_set.read_from_file(command_line_parser.test_filename);
	test_info.file_time = get_process_time_difference(test_info.file_time);
		
	if (training_set.dim() != test_set.dim())
		flush_exit(1, "Training and testing data have different dimensions.");
	
	testset_info = Tdataset_info(test_set, true);
	if ((test_set.is_classification_data() == false) or (testset_info.label_list.size() > 2))
		command_line_parser.display_roc_style_errors = false;

	
//--------- Test the decision functions -------------------------------------------------------------------

	svm_manager.load(training_set);
	svm_manager.test(test_set, command_line_parser.test_control, test_info);

	
//--------- Write obtained test results to file -----------------------------------------------------------

	test_info.file_time = get_process_time_difference(test_info.file_time);
	if (command_line_parser.result_filename != "")
	{
		fpresultwrite = open_file(command_line_parser.result_filename, "w");
		for(i=0; i<test_set.size(); i++)
		{
			predictions = svm_manager.get_predictions_for_test_sample(i);
			
			file_write(fpresultwrite, predictions[0], "%5.7g", "");
			for(j=1; j<predictions.size(); j++)
				file_write(fpresultwrite, predictions[j], ", %5.7g", "");
			file_write_eol(fpresultwrite);
		}
		close_file(fpresultwrite);
	}

	fplogwrite = open_file(command_line_parser.test_control.write_log_test_filename, "a");
	
	
	if (test_info.number_of_tasks == test_info.number_of_all_tasks)
		task_offset = 1;
	else
		task_offset = 0;
	
	for(j=0;j<test_info.train_val_info.size();j++)
	{
		test_info.train_val_info[j].val_time = test_info.test_time;
		test_info.train_val_info[j].write_to_file(fplogwrite);
		
		if (get_filetype(command_line_parser.test_filename) != NLA)
		{
			if (command_line_parser.display_roc_style_errors == false)
				flush_info(INFO_1,"\nTask %d: Test error %1.4f.",   j + task_offset, test_info.train_val_info[j].val_error);
			else
			{
				if (command_line_parser.test_control.vote_control.npl_class == -1)
				{
					detection_rate = 1.0 - test_info.train_val_info[j].pos_val_error;
					false_alarm_rate = test_info.train_val_info[j].neg_val_error;
				}
				else
				{
					detection_rate = 1.0 - test_info.train_val_info[j].neg_val_error;
					false_alarm_rate = test_info.train_val_info[j].pos_val_error;
				}
				flush_info(INFO_1,"\nTask %d: Detection Rate %1.4f.  False Alarm Rate %1.4f.", j + task_offset, detection_rate, false_alarm_rate);
			}
		}
	}
	close_file(fplogwrite);
	test_info.file_time = get_process_time_difference(test_info.file_time);


//----------------------- Final duties -----------------------------------------------------------------------------

	full_time = get_wall_time_difference(full_time);
	flush_info(INFO_1,"\n\n%4.2f seconds used to run svm-test.", full_time);
	flush_info(INFO_1,"\n%4.2f seconds used for testing.", test_info.test_time);
	flush_info(INFO_1,"\n%4.2f seconds used for file operations.", test_info.file_time);

	command_line_parser.copyright();

	flush_info(INFO_1,"\n\n");
}

File: bindings/java/junit/de/uni_stuttgart/isa/liquidsvm/SVMTest.java
// Copyright 2015-2017 Philipp Thomann
//
// This file is part of liquidSVM.
//
// liquidSVM is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// liquidSVM is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public License
// along with liquidSVM. If not, see <http://www.gnu.org/licenses/>.

package de.uni_stuttgart.isa.liquidsvm;

import static org.junit.Assert.*;

import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;

import de.uni_stuttgart.isa.liquidsvm.SVM.MC;
import static org.hamcrest.CoreMatchers.*;
import org.hamcrest.CoreMatchers;
import org.hamcrest.Description;
import org.hamcrest.TypeSafeMatcher;

/**
 * @author Philipp Thomann
 *
 */
public class SVMTest {
	
	private static final double ls_err = 5;
	private static final double mc_err = 0.05;
	private double[][] mc_tr;
	private double[] mc_tr_labs;
	private double[][] mc_ts;
	private double[] mc_ts_labs;
	private int mc_levels;
	
	private double[][] reg_tr;
	private double[] reg_tr_labs;
	private double[][] reg_ts;
	private double[] reg_ts_labs;

	@Before
	public void setUp() throws Exception {
		mc_tr = LiquidData.iris;
		mc_tr_labs = LiquidData.iris_labs;
		mc_ts = LiquidData.iris;
		mc_ts_labs = LiquidData.iris_labs;
		mc_levels = 3;
		
		reg_tr = LiquidData.trees;
		reg_tr_labs = LiquidData.trees_labs;
		reg_ts = LiquidData.trees;
		reg_ts_labs = LiquidData.trees_labs;
	}


	@Test
	public void testMC_AvA_hinge() {
		int n = mc_ts.length;
		int tasks = 1 + mc_levels * (mc_levels-1) / 2; // AvA
		
		SVM s = new SVM.MC(mc_tr,mc_tr_labs, theConfig());
		ResultAndErrors res = s.test(mc_ts,mc_ts_labs);
		assertEquals("lengths are equal", n, res.result.length);
		assertEquals("number of error columns", tasks, res.errors.length);
		double handErr = 0;
		for(int i=0; i<n; i++)
			handErr += (mc_ts_labs[i] == res.result[i][0]) ? 0:1;
		handErr /= (double)n;
		assertLT(handErr,mc_err);
		assertLT(res.errors[0][0],mc_err);
		assertApprox(res.errors[0][0], handErr);
	}
	
	@Test
	public void testMC_OvA_ls() {
		int n = mc_ts.length;
		int tasks = 1 + mc_levels ; // OvA
		
		SVM s = new SVM.MC(mc_tr,mc_tr_labs, "OvA", theConfig());
		ResultAndErrors res = s.test(mc_ts,mc_ts_labs);
		assertEquals("lengths are equal", n, res.result.length);
		assertEquals("number of error columns", tasks, res.errors.length);
		double handErr = 0;
		for(int i=0; i<n; i++)
			handErr += (mc_ts_labs[i] == res.result[i][0]) ? 0:1;
		handErr /= (double)n;
		assertLT(handErr,mc_err);
		assertLT(res.errors[0][0],mc_err);
		assertApprox(res.errors[0][0], handErr);
	}
	
	@Test
	public void testMC_OvA_hinge() {
		int n = mc_ts.length;
		int tasks = 1 + mc_levels ; // OvA
		
		SVM s = new SVM.MC(mc_tr,mc_tr_labs, "OvA_hi", theConfig());
		ResultAndErrors res = s.test(mc_ts,mc_ts_labs);
		assertEquals("lengths are equal", n, res.result.length);
		assertEquals("number of error columns", tasks, res.errors.length);
		double handErr = 0;
		for(int i=0; i<n; i++)
			handErr += (mc_ts_labs[i] == res.result[i][0]) ? 0:1;
		handErr /= (double)n;
		assertLT(handErr,mc_err);
		assertLT(res.errors[0][0],mc_err);
		assertApprox(res.errors[0][0], handErr);
	}
	
	@Test
	public void testMC_AvA_ls() {
		int n = mc_ts.length;
		int tasks = 1 + mc_levels * (mc_levels-1) / 2; // AvA
		
		SVM s = new SVM.MC(mc_tr,mc_tr_labs, "AvA_ls", theConfig());
		ResultAndErrors res = s.test(mc_ts,mc_ts_labs);
		assertEquals("lengths are equal", n, res.result.length);
		assertEquals("number of error columns", tasks, res.errors.length);
		double handErr = 0;
		for(int i=0; i<n; i++)
			handErr += (mc_ts_labs[i] == res.result[i][0]) ? 0:1;
		handErr /= (double)n;
		assertLT(handErr,mc_err);
		assertLT(res.errors[0][0],mc_err);
		assertApprox(res.errors[0][0], handErr);
	}
	
	@Test
	public void testLS() {
		int n = reg_ts.length;
		int tasks = 1;
		
		SVM s = new SVM.LS(reg_tr,reg_tr_labs, theConfig());
		ResultAndErrors res = s.test(reg_ts,reg_ts_labs);
		assertEquals("lengths are equal", n, res.result.length);
		assertEquals("number of error columns", tasks, res.errors.length);
		double handErr = 0;
		for(int i=0; i<n; i++){
			System.out.print(reg_ts_labs[i] +":"+ res.result[i][0]+", ");
			handErr += Math.pow(reg_ts_labs[i] - res.result[i][0], 2);
		}
		handErr /= (double)n;
		assertLT(handErr,ls_err);
		assertLT(res.errors[0][0],ls_err);
		assertApprox(res.errors[0][0], handErr);
	}


	//@Ignore @Test
	public void testParallel() throws InterruptedException {
		for(int i=0; i<1000; i++){
		Thread a = new Thread(new Runnable(){ public void run(){
				int n = reg_ts.length;
				int tasks = 1;
				
		SVM s = new SVM.LS(reg_tr,reg_tr_labs, theConfig().set("THREADS","2 0"));
		ResultAndErrors res = s.test(reg_ts,reg_ts_labs);
		assertEquals("lengths are equal", n, res.result.length);
		assertEquals("number of error columns", tasks, res.errors.length);
		double handErr = 0;
		for(int i=0; i<n; i++){
			System.out.print(reg_ts_labs[i] +":"+ res.result[i][0]+", ");
			handErr += Math.pow(reg_ts_labs[i] - res.result[i][0], 2);
		}
		handErr /= (double)n;
		assertLT(handErr,ls_err);
		assertLT(res.errors[0][0],ls_err);
		assertApprox(res.errors[0][0], handErr);
			}});
		a.run();
		Thread b = new Thread(new Runnable(){ public void run(){
			int n = reg_ts.length;
			int tasks = 1;
			
			SVM s = new SVM.LS(reg_tr,reg_tr_labs, theConfig().set("THREADS","2 2"));
			ResultAndErrors res = s.test(reg_ts,reg_ts_labs);
			assertEquals("lengths are equal", n, res.result.length);
			assertEquals("number of error columns", tasks, res.errors.length);
			double handErr = 0;
			for(int i=0; i<n; i++){
				System.out.print(reg_ts_labs[i] +":"+ res.result[i][0]+", ");
				handErr += Math.pow(reg_ts_labs[i] - res.result[i][0], 2);
			}
			handErr /= (double)n;
			assertLT(handErr,ls_err);
			assertLT(res.errors[0][0],ls_err);
			assertApprox(res.errors[0][0], handErr);
		}});
		b.run();
		a.join();
		b.join();
		}
	}


	private static Config theConfig() {
		return new Config().threads(1).display(0);
	}
	
	public static void assertLT(double x, double than){
		assertTrue(x+" should be < "+than, x<than);
	}
	
	public static void assertApprox(double a, double b){
		assertTrue(Math.abs(a-b)<0.001);
	}
	
}
